### PR TITLE
Allround improvements

### PR DESCRIPTION
**Pull request by jonasfreimuth**
* merge-summary-table-scripts
* simplify-deconvolution
* simplify-mutation-regression
* improve-index-rmd
* improve-qc-report-per-sample-rmd
* change-file-io-to-fread

### CODE REVIEW

File: scripts/creating_variants_summary_table.R [hunk 1]
```diff
@@ -1,16 +1,6 @@
 
-get_files <- function ( variants_dir){
```
Comment (rekado) [inline scripts/creating_variants_summary_table.R:2]: This is great!  I'm happy to see get_files gone!  This was a real problem, because it essentially circumvented input/output tracking.

File: snakefile.py [hunk 2]
```diff
@@ -672,7 +672,7 @@ def no_variant_vep(sample, lofreq_output):
 
 rule create_mutations_summary:
     input:
-        script = os.path.join(SCRIPTS_DIR, "creating_mutation_summary_table.R"),
+        script = os.path.join(SCRIPTS_DIR, "create_summary_table.R"),
```
Comment (rekado) [inline snakefile.py:689]: Nice!  Glad to see the duplication gone.

File: scripts/deconvolution_funs.R [hunk 3]
```diff
@@ -223,31 +223,38 @@ get_protein_mut <- function(vepfile) {
...
+  }) %>%
+
+    dplyr::bind_cols() %>%
+
+    magrittr::set_names(names(mutations_df)) %>%
+
+    # add "Others" col of all "0"s to indicate possible other variants not
+    # possessing any of the mutations
+    # TODO ensure this is how this column is supposed to work
+    mutate(Others = rep(0, length(mutations_vector))) %>%
```
Comment (rekado) [inline scripts/deconvolution_funs.R:165]: Is this really how it should behave?  All zeros?

The original comment said this:

> "Others" means that the particular mutation is not found and the mutation site could be mutated otherwise or not at all

I suppose in the old code Others would always be zero, because it loops over `names(mutations.df)`, which I assume never contains "Others".

Frankly, all of this is really hard to follow for me, and the lack of documentation or tests makes it really hard to know what invariants there are.

File: scripts/deconvolution_funs.R [hunk 4]
```diff
@@ -111,76 +112,111 @@ get_protein_mut <- function(vepfile) {
   # reading in whole vep txt output
   # you should include in the vep script to parse out the #
   # in the beginning of the line or include that step here.
-  vepfile.df <- read.table(vepfile, sep = ",", header = TRUE)
+  vepfile_df <- read.table(vepfile, sep = ",", header = TRUE)
```
Comment (rekado) [inline scripts/deconvolution_funs.R]: I'm having a hard time following the changes below due to simultaneous renaming, abbreviations, and changed regular expressions.

File: scripts/deconvolution.R [hunk 5]
```diff
@@ -190,55 +190,72 @@ if (execute_deconvolution) {
   # for the deconvolution to work we need the "wild type" frequencies too.
   # The matrix from above got mirrored, wild type mutations are simulated the
   # following: e.g. T210I (mutation) -> T210T ("wild type")
-  msig_simple <- create_sig_matrix(mutations_vec, mutation_sheet) %>%
+  msig_simple <- create_sig_matrix(mutations_vec, mutation_sheet)
```
Comment (rekado) [inline scripts/deconvolution.R:181]: There are too many changes in commit 656217547482dc02257691fc933ca78600256ad9, so I cannot convince myself that the changes are equivalent.  I'll just trust that you know what you've done here.

File: snakefile.py [hunk 6]
```diff
@@ -595,14 +597,18 @@ def no_variant_vep(sample, lofreq_output):
     shell:
         """
         {RSCRIPT_EXEC} {input.script} \
+        "{input.deconvolution_functions}" \
         "{wildcards.sample}" \
-        "{OUTPUT_DIR}" \
+        "{MUTATION_SHEET_CSV}" \
+        "{SAMPLE_SHEET_CSV}" \
```
Comment (rekado) [inline snakefile.py:603]: This is not new in this commit (6a365b1321d924f5870987a6e8f60825d32ebcac), but I'll mention it anyway: these CSV files should also be listed in the inputs of this rule to ensure that they are made available when this rule is executed.  This is especially important for distributed execution (e.g. on a cluster or on cloud VMs) where the inputs must be copied to the machine first.

File: scripts/pred_mutation_increase.R [hunk 7]
```diff
@@ -22,63 +22,45 @@ parsing_mutation_plot_data <- function ( mutation_plot_data ){
...
+  # run a linear model of each mutation col (all cols except dates)
+  # to get the change of each mutations abundance over all dates
+  all_lm_res <- lapply(
+    mutations,
+    function(mutation) {
+      mutation_col <- mutations_df[[mutation]]
+
+      # only do regression if >= 20% of dates have detected mutations
+      # (if not, this will return NULL)
+      if (sum(mutation_col > 0) >= 0.2 * length(dates)) {
```
Comment (rekado) [inline scripts/pred_mutation_increase.R:27]: Due to this single-armed `if` this is a partial function.  In the `else` case there is no defined return value.  Is this intentional?  Should this instead be filtered first and then the total function be applied unconditionally?

(849b6474def07cb5a840918d9e57b9f26eba5f5d)

File: scripts/report_scripts/index.Rmd [hunk 8]
```diff
@@ -478,10 +478,16 @@ mutations_sig <- filter_lm_res_top20(mutations_sig_unfiltered, 0.05)
...
-      sigmuts_df == str_split(mutation, str_glue("{mut_str_sep}{mut_str_sep}"))[[1]][2]
-    ), "SigMut", ""))
+    mutate(mut_str = str_extract(
+      mutation,
+      str_glue("(?<={mut_str_sep}{mut_str_sep})[[:alnum:]]+"))) %>%
+    mutate(
+      sigflag = ifelse(
+        mut_str %in%  unlist(sigmuts_df),
+        "SigMut",
+        "NoSigMut"
```
Comment (rekado) [inline scripts/report_scripts/index.Rmd]: I don't see where NoSigMut comes from compared with the code before this commit (937c21a813eddb150101be70ba3fc78a97d8c3ae).

File: scripts/report_scripts/index.Rmd [hunk 9]
```diff
@@ -639,15 +652,16 @@ melted <- reshape2::melt(filtered_approved_mut_plot,
...
-        variable, str_glue("{mut_str_sep}{mut_str_sep}")
-      )[[1]][2]),
-      "SigMut", NULL
-    )
-  )
+      mut_str %in%  unlist(sigmuts_df),
+      "SigMut",
+      "NoSigMut"
+  )) %>%
+  dplyr::select(-mut_str)
```
Comment (rekado) [inline scripts/report_scripts/index.Rmd:682]: This pattern has been repeated a couple of times here.  Can we move this transformation to a procedure instead of repeating it literally?

**Review by rekado — APPROVED** [review]
This looks all pretty good to me; a significant improvement!
I wasn't able to follow all changes, because some commits were quite dense.  I added a few comments, but overall I'd say: go for it!

Thank you!

File: scripts/pred_mutation_increase.R (hunk 7 above)
Comment (jonasfreimuth) [reply scripts/pred_mutation_increase.R:27]: Yes this is intentional. In the alternative case the function will return `NULL` (invisibly), which will be dropped during `bind_rows`

File: scripts/report_scripts/index.Rmd (hunk 8 above)
Comment (jonasfreimuth) [reply scripts/report_scripts/index.Rmd]: This is what I meant by "unifying" the creation of this flag. Previously the alternative values were just an empty string or `NULL` which would be filled with "nosigmut". Now it is always "NoSigMut"  which is easiery to read than nosigmut and more expressive than "".

File: scripts/report_scripts/index.Rmd (hunk 9 above)
Comment (jonasfreimuth) [reply scripts/report_scripts/index.Rmd:682]: I will do that the next time I touch that notebook.

File: snakefile.py (hunk 6 above)
Comment (jonasfreimuth) [reply snakefile.py:603]: I will also do that the next time I touch the snakefile. I would also like to go through the scripts and enforce the input output rule, removing loading of unspecified files in scripts.

File: scripts/deconvolution_funs.R (hunk 3 above)
Comment (jonasfreimuth) [reply scripts/deconvolution_funs.R:165]: Yes and that is in line with how simulateOthers worked, because then in the logical inverse matrix it becomes all 1, simulating that variant having the NULL mutations at the loci where other variants have actual mutations.

File: scripts/pred_mutation_increase.R (hunk 7 above)
Comment (rekado) [reply scripts/pred_mutation_increase.R:27]: Ah, good, that makes sense.  (In Scheme that would be literally the unspecified value, which is distinct from the equivalent of NULL.)

File: scripts/report_scripts/index.Rmd (hunk 8 above)
Comment (rekado) [reply scripts/report_scripts/index.Rmd]: Thanks for clarifying. This is what I guessed, but wasn't sure about.

File: scripts/deconvolution_funs.R (hunk 4 above)
Comment (jonasfreimuth) [reply scripts/deconvolution_funs.R]: In the commit message for that commit I now added some explanations what exatcly the new code does vs the old...
